### PR TITLE
[tune] Only keep cached actors if search has not ended

### LINKED ISSUE #31883: [Ray 2.3 release] `cluster_tune_scale_up_down` release test fails
https://buildkite.com/ray-project/release-tests-branch/builds/1303#0185d157-4ce0-43da-a273-478fd7fbf07e

```
2023-01-20 15:32:36,885	INFO tune.py:782 -- Total run time: 1806.15 seconds (1805.96 seconds for the tuning loop).
--
  | Traceback (most recent call last):
  | File "workloads/tune_scale_up_down.py", line 57, in <module>
  | main()
  | File "workloads/tune_scale_up_down.py", line 53, in main
  | assert node_counter.node_counts[-1] == 1, "Cluster didn't scale down to 1 node."
  | AssertionError: Cluster didn't scale down to 1 node.
```

### PR DESCRIPTION
**Pull request by krfricke**
Signed-off-by: Kai Fricke <kai@anyscale.com>

## Why are these changes needed?

We currently keep one actor cached if no other trials have been staged to prevent us removing actors when we may need them in one of the next iterations. However, when the search algorithm won't produce new trials anymore, we keep this actor needlessly. This can keep resources occupied needlessly and prevent downscaling, as caught in #31883.

This PR passes another flag to the trial executor cleanup method indicating if new trials are to be expected. If not, we can cleanup unneeded actors even if no further trials are staged.

This behavior is tested in the `cluster_tune_scale_up_down` release test. We won't add a unit test for this as this would effectively mimic the release test. Instead, we can add proper actor reuse testing once we made more progress with the execution refactor.

## Related issue number

Closes #31883

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by xwjiang2010** [comment]
Thanks!
The test is written very well to catch it!
Since we are at it, do yo mind adding a small description on top of `tune_scale_up_down.py`? Like the ones we have in tune scalability envelope.

**Review by xwjiang2010 — APPROVED** [review]

**Comment by krfricke** [comment]
Updated the description of the test!

**Comment by krfricke** [comment]
Release test passed: https://buildkite.com/ray-project/release-tests-pr/builds/26587

Waiting for rest of CI to pass before merging.

**Pull request merged**
